### PR TITLE
feat: otel instrumentation of wundernode and server

### PR DESCRIPTION
**Pull request by StarpTech**
image

This PR adds OpenTelemetry instrumentation to the WunderNode and Server. Due to an OTEL dependency, I was forced to upgrade go to 1.19

Docs: 
- https://wundergraph-docs-git-dustin-eng-1548-instru-4adca8-wunder-graph.vercel.app/docs/guides/monitoring-and-observability
- https://wundergraph-docs-git-dustin-eng-1548-instru-4adca8-wunder-graph.vercel.app/docs/wundergraph-config-ts-reference/configure-wundernode-options

## Done

- [X] Instrument WunderNode router
- [X] Instrument WunderNode Webhooks proxy
- [X] Instrument Hooks Client
- [X] Instrument Hooks server router
- [X] Instrument Operation Client requests
- [x] Add documentation, API reference
- [x] Ensure that HTTP errors are handled the same in all instrumentation
- [x] Span attribute alignments (naming)
- [X] Add official example
- [X] Zero overhead when tracing is disabled
- [X] Removed deprecated `/main` entrypoint.
- [ ] Instrument ORM requests (Not planned in this PR)

## TODO

- [x] Add more tests
- [x] Validate against public cloud otel collector
- [x] Add documentation, API reference
- [x] Ensure that HTTP errors are handled the same in all instrumentation
- [x] Span attribute alignments (naming)

## Misc

- Internal client was skipped because we deprecated it months ago and will remove it in June.
- Consider adding events in the future

#### Checklist

- [ ] run `make test`
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: cli/commands/root.go [hunk 1]
```diff
@@ -268,6 +270,12 @@ func init() {
 	_, isTelemetryDebugEnabled := os.LookupEnv("WG_TELEMETRY_DEBUG")
 	telemetryAnonymousID := os.Getenv("WG_TELEMETRY_ANONYMOUS_ID")
 
+	if otelBatchTimeoutEnv, ok := os.LookupEnv("WG_OTEL_BATCH_TIMEOUT_MS"); ok {
+		if ms, err := strconv.Atoi(otelBatchTimeoutEnv); err == nil {
+			otelBatchTimeout = time.Duration(ms) * time.Millisecond
+		}
```
Comment (Aenimus) [inline cli/commands/root.go:276]: do nothing with the err?

Comment (StarpTech) [reply cli/commands/root.go:279]: ty, fixed.

File: examples/open-telemetry/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+version: '3.7'
```
Comment (ggeijteman-wg) [inline examples/open-telemetry/docker-compose.yml:1]: Nit: unless we need `3.7` for anything specific, the latest version is `3.8`

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -9,7 +9,7 @@ This guide assumes you have already installed the following software:
 - make (should be installed on all linux / IOS systems)
 - [protoc](https://grpc.io/docs/protoc-installation/) (3.15.8)
   - curl -LO "https://github.com/protocolbuffers/protobuf/releases/download/v21.5/protoc-21.5-linux-x86_64.zip"
-- golang `>= 1.18`
+- golang `>= 1.19`
```
Comment (ggeijteman-wg) [inline CONTRIBUTING.md:12]: With `go1.20` out, why do we only jump to 19?

File: packages/sdk/src/configure/options.ts [hunk 4]
```diff
@@ -17,6 +17,11 @@ export enum WgEnv {
 	HttpProxyUrl = 'WG_HTTP_PROXY',
 	PrometheusEnabled = 'WG_PROMETHEUS_ENABLED',
 	PrometheusPort = 'WG_PROMETHEUS_PORT',
+	OtelEnabled = 'WG_OTEL_ENABLED',
```
Comment (ggeijteman-wg) [inline packages/sdk/src/configure/options.ts:20]: Nit: Is there a matching PR to add these ENV vars to the cloud?

File: pkg/apihandler/apihandler.go [hunk 5]
```diff
@@ -192,44 +198,10 @@ func (r *Builder) BuildAndMountApiHandler(ctx context.Context, router *mux.Route
 		r.planConfig.DataSources = append(r.planConfig.DataSources, dataSource)
 	}
 
-	// limiter := rate.NewLimiter(rate.Every(time.Second), 10)
-
 	r.log.Debug("configuring API",
 		zap.Int("numOfOperations", len(api.Operations)),
 	)
 
-	// Redirect from old-style URL, for temporary backwards compatibility
```
Comment (ggeijteman-wg) [inline pkg/apihandler/apihandler.go:201]: Nit: things that are not related to this feature, maybe should be a separate PR. I like getting rid of stuff we don't need though 👍

File: pkg/node/node.go [hunk 6]
```diff
@@ -281,18 +320,27 @@ func (n *Node) Close() error {
...
+
 	if n.internalServer != nil {
 		if err := n.internalServer.Close(); err != nil {
 			return err
 		}
 		n.internalServer = nil
 	}
+
+	if n.tracer != nil {
+		if err := n.tracer.Shutdown(context.Background()); err != nil {
```
Comment (ggeijteman-wg) [inline pkg/node/node.go:339]: I'm not sure I understand, why aren't we using the available `ctx` here?

**Review by ggeijteman-wg — APPROVED** [review]
Great work 👍, concise code, well thought of test cases and commented well.

File: packages/sdk/src/configure/options.ts (hunk 4 above)
Comment (StarpTech) [reply packages/sdk/src/configure/options.ts:20]: Not yet, but that is a very good point @Aenimus. This must be set.

File: pkg/node/node.go (hunk 6 above)
Comment (StarpTech) [reply pkg/node/node.go:339]: We have no context here. `Close` is the forcefully equivalent of `.Shutdown`

File: packages/sdk/src/configure/options.ts (hunk 4 above)
Comment (Aenimus) [reply packages/sdk/src/configure/options.ts:20]: Did you mean to ping me? :D

File: CONTRIBUTING.md (hunk 3 above)
Comment (StarpTech) [reply CONTRIBUTING.md:12]: Because the minimum version I needed was 1.19.

File: examples/open-telemetry/docker-compose.yml (hunk 2 above)
Comment (StarpTech) [reply examples/open-telemetry/docker-compose.yml:1]: I think I can use `3` to pick up minors.

File: CONTRIBUTING.md (hunk 3 above)
Comment (Aenimus) [reply CONTRIBUTING.md:12]: Shall we use 1.20? We're using that where we can, after all.

File: packages/sdk/src/configure/options.ts (hunk 4 above)
Comment (StarpTech) [reply packages/sdk/src/configure/options.ts:20]: Sry, it was intended for @fiam

Comment (Aenimus) [reply packages/sdk/src/configure/options.ts:20]: Dustin are you ok? Fiam is Alberto :D

Comment (Aenimus) [reply packages/sdk/src/configure/options.ts:20]: Ahh, you want Alberto to set it? I thought you wanted to ping Gerben :D I was worried

File: examples/open-telemetry/docker-compose.yml (hunk 2 above)
Comment (StarpTech) [reply examples/open-telemetry/docker-compose.yml]: Done

File: packages/sdk/src/configure/options.ts (hunk 4 above)
Comment (StarpTech) [reply packages/sdk/src/configure/options.ts:20]: Will close it and comment it on @fiam PR.

File: cli/commands/root.go [hunk 7]
```diff
@@ -41,6 +42,7 @@ var (
 	cmdDurationMetric     telemetry.DurationMetric
 	zapLogLevel           zapcore.Level
 	_wunderGraphDirConfig string
+	otelBatchTimeout      time.Duration
```
Comment (jensneuse) [inline cli/commands/root.go:45]: can you add a comment what this does, how it works, what are reasonable defaults?

File: docs-website/src/pages/docs/guides/monitoring-and-observability.md [hunk 8]
```diff
@@ -16,22 +16,79 @@ application's performance. The following metrics are available:
...
+If you are not familiar with OpenTelemetry, you can read more about it [here](https://opentelemetry.io/).
+The OpenTelemetry integration is disabled by default. To enable it, set `enabled` to `true` and optionally configure the endpoint and sampling rate:
+
+```typescript
+// wundergraph.config.ts
+
+configureWunderGraphApplication({
+  options: {
+    openTelemetry: {
+      enabled: true, // Disabled by default
```
Comment (jensneuse) [inline docs-website/src/pages/docs/guides/monitoring-and-observability.md:56]: rename to "enable"?

File: docs-website/src/pages/docs/guides/monitoring-and-observability.md [hunk 9]
```diff
@@ -16,22 +16,79 @@ application's performance. The following metrics are available:
...
+      sampler: 1, // Defaults to 1 (every trace request). Samples a given fraction of traces. Must be a value between 0 and 1.
```
Comment (jensneuse) [inline docs-website/src/pages/docs/guides/monitoring-and-observability.md:57]: Please add a section below to explain what the sampler value does, what are reasonable values, what should users consider.

File: packages/sdk/src/server/plugins/hooks.ts [hunk 10]
```diff
@@ -56,178 +66,254 @@ const FastifyHooksPlugin: FastifyPluginAsync<FastifyHooksOptions> = async (fasti
...
-	});
-
-	// wsTransport
-	if (config.global?.wsTransport?.onConnectionInit) {
-		fastify.post<{
+	fastify.post<
+		{
 			Body: {
-				dataSourceId: string;
 				request: WunderGraphRequest;
```
Comment (jensneuse) [inline packages/sdk/src/server/plugins/hooks.ts:174]: we could move this into an interface definition

File: packages/sdk/src/server/plugins/hooks.ts [hunk 11]
```diff
@@ -56,178 +66,254 @@ const FastifyHooksPlugin: FastifyPluginAsync<FastifyHooksOptions> = async (fasti
...
-				hook: 'onConnectionInit',
-			});
-		});
+		}
+	);
+
+	fastify.post<
+		{
+			Body: {
+				response: WunderGraphResponse;
```
Comment (jensneuse) [inline packages/sdk/src/server/plugins/hooks.ts:226]: and re-use it here

File: pkg/apihandler/apihandler.go [hunk 12]
```diff
@@ -922,7 +906,10 @@ type QueryHandler struct {
 
 func (h *QueryHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	requestLogger := h.log.With(logging.WithRequestIDFromContext(r.Context()))
-	r = setOperationMetaData(r, h.operation)
+	r = operation.SetOperationMetaData(r, h.operation)
```
Comment (jensneuse) [inline pkg/apihandler/apihandler.go:909]: would it make sense to merge line 908, 909 and 912 into one single function that returns the logger, the curried request and handles tracing?

File: pkg/node/node.go [hunk 13]
```diff
@@ -158,6 +162,17 @@ func WithGlobalRateLimit(requests int, perDuration time.Duration) Option {
 	}
 }
 
+// WithTraceBatchTimeout sets the timeout for the trace batch exporter
+// If the timeout is less or equal than 0, the default value of 5 seconds will be used
```
Comment (jensneuse) [inline pkg/node/node.go:166]: I would appreciate more info here on what exactly to configure here.

File: pkg/node/node.go (hunk 6 above)
Comment (jensneuse) [inline pkg/node/node.go:339]: Would it make sense to "close" with a context and a small timeout? This way we could ensure that the tracer flushes before we shut it down. With ctx background we might not wait long enough maybe?

**Review by jensneuse — APPROVED** [review]
a few minor things, overall looks very good, you can merge or address some of the minor feedback, I'm approving already

File: docs-website/src/pages/docs/guides/monitoring-and-observability.md (hunk 8 above)
Comment (StarpTech) [reply docs-website/src/pages/docs/guides/monitoring-and-observability.md:56]: Prometheus uses `enabled` too. Should we rename all or stick with `enabled`?

File: pkg/apihandler/apihandler.go (hunk 12 above)
Comment (StarpTech) [reply pkg/apihandler/apihandler.go:909]: It doesn't feel right to me and is like a pre-opt.

File: packages/sdk/src/server/plugins/hooks.ts (hunk 10 above)
Comment (StarpTech) [reply packages/sdk/src/server/plugins/hooks.ts:174]: I don't know there is a very small overlap. I'd like to keep it.

File: pkg/node/node.go (hunk 13 above)
Comment (StarpTech) [reply pkg/node/node.go:166]: Done

File: docs-website/src/pages/docs/guides/monitoring-and-observability.md (hunk 9 above)
Comment (StarpTech) [reply docs-website/src/pages/docs/guides/monitoring-and-observability.md:57]: Done

File: cli/commands/root.go (hunk 7 above)
Comment (StarpTech) [reply cli/commands/root.go:45]: Done

File: pkg/node/node.go (hunk 6 above)
Comment (StarpTech) [reply pkg/node/node.go:339]: This is the `Close()` method which is only used to forcefully exit the server. We use `.Shutdown` in production which calls `flush` before.

**Review by jensneuse — APPROVED** [review]
